clarity_ext/service/artifact_service.py

```diff
@@ -106,3 +118,45 @@ def all_shared_result_files(self):
     def update_artifacts(self, update_queue):
         response = self.step_repository.update_artifacts(update_queue)
         return response
+
+    def parent_input_artifacts(self):
+        """
+        Returns a dictionary of all parent input artifacts, indexed by process ID.
+
+        Details:
+        In your current step (CS), you will have input and output artifacts like this:
+            [CS-I1  ->  CS-O1]
+            [CS-I2  ->  CS-O2]
+            ...
+
+        These artifacts may be coming from a parent step (PS):
+            [PS-I1  ->  PS-O1]  -> [CS-I1   -> CS-O2]
+            ...
+
+        This method will fetch all of the input artifacts based on all of your output artifacts in one call
+        and index them by their respective process id.
+        """
+
+        # We will need the input artifacts from the previous step
+        parent_processes = set([artifact.input.parent_process for artifact in self.all_output_artifacts()])
+
+        for process in parent_processes:
```

Comment:
So a `Process` is one step in a workflow protocol. There is some confusion about these sometimes being called steps and sometimes processes, but the API endpoint is `/processes`.

Yes, I only fetch for the parent process in this case, but future calls might try to fetch the entire tree. I think that you might be right that that would cause circular fetches, but I don't think it's a concern in this commit.